Add Python 3.11-dev to Github actions

### Current problem

Seems like [tomllib](https://peps.python.org/pep-0680/) has been accepted and merged https://github.com/python/cpython/pull/31498 but not yet released to any of the images available on Github.

[Next release](https://peps.python.org/pep-0664/) is expected on `2022-04-05` so we'll need to wait a bit.

### Desired solution

Add 3.11-dev when github actions are up to date and released.

### Additional context

https://github.com/PyCQA/pylint/issues/5919#issuecomment-1067728097

I can say it currently doesn't work with latest 3.11.0a7:

```
ImportError while importing test module '/builds/hackancuba/asgi-signing-middleware/asgi_signing_middleware/tests/tests_cookie_starlite.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/local/lib/python3.11/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
asgi_signing_middleware/tests/tests_cookie_starlite.py:26: in <module>
    from starlite import Request
.cache/poetry/virtualenvs/asgi-signing-middleware-4gCN4nBn-py3.11/lib/python3.11/site-packages/starlite/__init__.py:3: in <module>
    from .app import Starlite
.cache/poetry/virtualenvs/asgi-signing-middleware-4gCN4nBn-py3.11/lib/python3.11/site-packages/starlite/app.py:19: in <module>
    from starlite.asgi import StarliteASGIRouter
.cache/poetry/virtualenvs/asgi-signing-middleware-4gCN4nBn-py3.11/lib/python3.11/site-packages/starlite/asgi.py:10: in <module>
    from starlite.routes import ASGIRoute, HTTPRoute
.cache/poetry/virtualenvs/asgi-signing-middleware-4gCN4nBn-py3.11/lib/python3.11/site-packages/starlite/routes.py:18: in <module>
    from starlite.controller import Controller
.cache/poetry/virtualenvs/asgi-signing-middleware-4gCN4nBn-py3.11/lib/python3.11/site-packages/starlite/controller.py:6: in <module>
    from starlite.handlers import BaseRouteHandler
.cache/poetry/virtualenvs/asgi-signing-middleware-4gCN4nBn-py3.11/lib/python3.11/site-packages/starlite/handlers/__init__.py:1: in <module>
    from .asgi import ASGIRouteHandler, asgi
.cache/poetry/virtualenvs/asgi-signing-middleware-4gCN4nBn-py3.11/lib/python3.11/site-packages/starlite/handlers/asgi.py:8: in <module>
    from starlite.handlers.base import BaseRouteHandler
.cache/poetry/virtualenvs/asgi-signing-middleware-4gCN4nBn-py3.11/lib/python3.11/site-packages/starlite/handlers/base.py:21: in <module>
    from starlite.provide import Provide
.cache/poetry/virtualenvs/asgi-signing-middleware-4gCN4nBn-py3.11/lib/python3.11/site-packages/starlite/provide.py:10: in <module>
    from starlite.signature import SignatureModel
.cache/poetry/virtualenvs/asgi-signing-middleware-4gCN4nBn-py3.11/lib/python3.11/site-packages/starlite/signature.py:8: in <module>
    from pydantic_factories import ModelFactory
.cache/poetry/virtualenvs/asgi-signing-middleware-4gCN4nBn-py3.11/lib/python3.11/site-packages/pydantic_factories/__init__.py:3: in <module>
    from .extensions import (
.cache/poetry/virtualenvs/asgi-signing-middleware-4gCN4nBn-py3.11/lib/python3.11/site-packages/pydantic_factories/extensions/__init__.py:2: in <module>
    from .beanie_odm import BeanieDocumentFactory, BeaniePersistenceHandler
.cache/poetry/virtualenvs/asgi-signing-middleware-4gCN4nBn-py3.11/lib/python3.11/site-packages/pydantic_factories/extensions/beanie_odm.py:6: in <module>
    from pydantic_factories.factory import ModelFactory
.cache/poetry/virtualenvs/asgi-signing-middleware-4gCN4nBn-py3.11/lib/python3.11/site-packages/pydantic_factories/factory.py:93: in <module>
    from pydantic_factories.constraints.strings import (
.cache/poetry/virtualenvs/asgi-signing-middleware-4gCN4nBn-py3.11/lib/python3.11/site-packages/pydantic_factories/constraints/strings.py:3: in <module>
    from exrex import getone  # pylint: disable=import-error
.cache/poetry/virtualenvs/asgi-signing-middleware-4gCN4nBn-py3.11/bin/exrex.py:25: in <module>
    from re import sre_parse, U
E   ImportError: cannot import name 'sre_parse' from 're' (/usr/local/lib/python3.11/re/__init__.py)
```

:point_right: [Full trace and context](https://gitlab.com/hackancuba/asgi-signing-middleware/-/jobs/2338927237)

--- a/.github/workflows/changelog.yml
+++ b/.github/workflows/changelog.yml
@@ -6,7 +6,7 @@ on:
 
 env:
   # Also change CACHE_VERSION in the other workflows
-  CACHE_VERSION: 27
+  CACHE_VERSION: 28
   DEFAULT_PYTHON: "3.10"
 
 jobs:

--- a/.github/workflows/changelog.yml
+++ b/.github/workflows/changelog.yml
@@ -7,7 +7,7 @@ on:
 env:
   # Also change CACHE_VERSION in the other workflows
   CACHE_VERSION: 29
-  DEFAULT_PYTHON: "3.10"
+  DEFAULT_PYTHON: "3.11-dev"
 
 jobs:
   check-changelog:

--- a/.github/workflows/changelog.yml
+++ b/.github/workflows/changelog.yml
@@ -6,7 +6,7 @@ on:
 
 env:
   # Also change CACHE_VERSION in the other workflows
-  CACHE_VERSION: 28
+  CACHE_VERSION: 29
   DEFAULT_PYTHON: "3.10"
 
 jobs:

--- a/.github/workflows/checks.yaml
+++ b/.github/workflows/checks.yaml
@@ -8,7 +8,7 @@ on:
   pull_request: ~
 
 env:
-  CACHE_VERSION: 28
+  CACHE_VERSION: 29
   DEFAULT_PYTHON: "3.10"
   PRE_COMMIT_CACHE: ~/.cache/pre-commit
 

--- a/.github/workflows/checks.yaml
+++ b/.github/workflows/checks.yaml
@@ -8,7 +8,7 @@ on:
   pull_request: ~
 
 env:
-  CACHE_VERSION: 27
+  CACHE_VERSION: 28
   DEFAULT_PYTHON: "3.10"
   PRE_COMMIT_CACHE: ~/.cache/pre-commit
 

--- a/.github/workflows/checks.yaml
+++ b/.github/workflows/checks.yaml
@@ -9,7 +9,7 @@ on:
 
 env:
   CACHE_VERSION: 29
-  DEFAULT_PYTHON: "3.10"
+  DEFAULT_PYTHON: "3.11-dev"
   PRE_COMMIT_CACHE: ~/.cache/pre-commit
 
 concurrency:

--- a/.github/workflows/primer-test.yaml
+++ b/.github/workflows/primer-test.yaml
@@ -13,7 +13,7 @@ on:
       - ".github/workflows/primer-test.yaml"
 
 env:
-  CACHE_VERSION: 27
+  CACHE_VERSION: 28
 
 concurrency:
   group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}

--- a/.github/workflows/primer-test.yaml
+++ b/.github/workflows/primer-test.yaml
@@ -13,7 +13,7 @@ on:
       - ".github/workflows/primer-test.yaml"
 
 env:
-  CACHE_VERSION: 28
+  CACHE_VERSION: 29
 
 concurrency:
   group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}

--- a/.github/workflows/primer-test.yaml
+++ b/.github/workflows/primer-test.yaml
@@ -26,7 +26,7 @@ jobs:
     timeout-minutes: 5
     strategy:
       matrix:
-        python-version: [3.8, 3.9, "3.10"]
+        python-version: [3.8, 3.9, "3.10", "3.11-dev"]
     outputs:
       python-key: ${{ steps.generate-python-key.outputs.key }}
     steps:
@@ -68,7 +68,7 @@ jobs:
     needs: prepare-tests-linux
     strategy:
       matrix:
-        python-version: [3.8, 3.9, "3.10"]
+        python-version: [3.8, 3.9, "3.10", "3.11-dev"]
     steps:
       - name: Check out code from GitHub
         uses: actions/checkout@v3.0.2
@@ -103,7 +103,7 @@ jobs:
     needs: prepare-tests-linux
     strategy:
       matrix:
-        python-version: [3.8, 3.9, "3.10"]
+        python-version: [3.8, 3.9, "3.10", "3.11-dev"]
     steps:
       - name: Check out code from GitHub
         uses: actions/checkout@v3.0.2

--- a/.github/workflows/primer_comment.yaml
+++ b/.github/workflows/primer_comment.yaml
@@ -37,7 +37,7 @@ jobs:
         id: python
         uses: actions/setup-python@v4.2.0
         with:
-          python-version: "3.10"
+          python-version: "3.11-dev"
 
       # Restore cached Python environment
       - name: Restore Python virtual environment

--- a/.github/workflows/primer_comment.yaml
+++ b/.github/workflows/primer_comment.yaml
@@ -14,7 +14,7 @@ on:
 
 env:
   # This needs to be the SAME as in the Main and PR job
-  CACHE_VERSION: 27
+  CACHE_VERSION: 28
 
 permissions:
   contents: read

--- a/.github/workflows/primer_comment.yaml
+++ b/.github/workflows/primer_comment.yaml
@@ -14,7 +14,7 @@ on:
 
 env:
   # This needs to be the SAME as in the Main and PR job
-  CACHE_VERSION: 28
+  CACHE_VERSION: 29
 
 permissions:
   contents: read

--- a/.github/workflows/primer_run_main.yaml
+++ b/.github/workflows/primer_run_main.yaml
@@ -25,7 +25,7 @@ jobs:
     timeout-minutes: 60
     strategy:
       matrix:
-        python-version: ["3.7", "3.10"]
+        python-version: ["3.7", "3.11-dev"]
     steps:
       - name: Check out code from GitHub
         uses: actions/checkout@v3.0.2

--- a/.github/workflows/primer_run_main.yaml
+++ b/.github/workflows/primer_run_main.yaml
@@ -16,7 +16,7 @@ concurrency:
 
 env:
   # This needs to be the SAME as in the PR and comment job
-  CACHE_VERSION: 27
+  CACHE_VERSION: 28
 
 jobs:
   run-primer:

--- a/.github/workflows/primer_run_main.yaml
+++ b/.github/workflows/primer_run_main.yaml
@@ -34,7 +34,6 @@ jobs:
         uses: actions/setup-python@v4.2.0
         with:
           python-version: ${{ matrix.python-version }}
-
       - name: Get latest astroid commit
         id: get-astroid-sha
         run: |
@@ -88,6 +87,11 @@ jobs:
           path: tests/.pylint_primer_tests/commit_string.txt
 
       # Run primer
+      - if: "contains(matrix.python-version, '-dev')"
+        # TODO: 2.16 Remove this when dill release their next version
+        run: |
+          . venv/bin/activate
+          pip install git+https://github.com/uqfoundation/dill.git@25a7e450ed76c7a0820834a3a91134476b1b8253
       - name: Run pylint primer
         run: |
           . venv/bin/activate

--- a/.github/workflows/primer_run_main.yaml
+++ b/.github/workflows/primer_run_main.yaml
@@ -16,7 +16,7 @@ concurrency:
 
 env:
   # This needs to be the SAME as in the PR and comment job
-  CACHE_VERSION: 28
+  CACHE_VERSION: 29
 
 jobs:
   run-primer:

--- a/.github/workflows/primer_run_main.yaml
+++ b/.github/workflows/primer_run_main.yaml
@@ -34,6 +34,7 @@ jobs:
         uses: actions/setup-python@v4.2.0
         with:
           python-version: ${{ matrix.python-version }}
+
       - name: Get latest astroid commit
         id: get-astroid-sha
         run: |
@@ -87,11 +88,6 @@ jobs:
           path: tests/.pylint_primer_tests/commit_string.txt
 
       # Run primer
-      - if: "contains(matrix.python-version, '-dev')"
-        # TODO: 2.16 Remove this when dill release their next version
-        run: |
-          . venv/bin/activate
-          pip install git+https://github.com/uqfoundation/dill.git@25a7e450ed76c7a0820834a3a91134476b1b8253
       - name: Run pylint primer
         run: |
           . venv/bin/activate

--- a/.github/workflows/primer_run_pr.yaml
+++ b/.github/workflows/primer_run_pr.yaml
@@ -148,11 +148,6 @@ jobs:
           git pull origin ${{ steps.download-main-run.outputs.result }} --no-edit --no-commit --no-rebase
 
       # Run primer
-      - if: "contains(matrix.python-version, '-dev')"
-        # TODO: 2.16 Remove this when dill release their next version
-        run: |
-          . venv/bin/activate
-          pip install git+https://github.com/uqfoundation/dill.git@25a7e450ed76c7a0820834a3a91134476b1b8253
       - name: Run pylint primer
         run: |
           . venv/bin/activate

--- a/.github/workflows/primer_run_pr.yaml
+++ b/.github/workflows/primer_run_pr.yaml
@@ -148,6 +148,11 @@ jobs:
           git pull origin ${{ steps.download-main-run.outputs.result }} --no-edit --no-commit --no-rebase
 
       # Run primer
+      - if: "contains(matrix.python-version, '-dev')"
+        # TODO: 2.16 Remove this when dill release their next version
+        run: |
+          . venv/bin/activate
+          pip install git+https://github.com/uqfoundation/dill.git@25a7e450ed76c7a0820834a3a91134476b1b8253
       - name: Run pylint primer
         run: |
           . venv/bin/activate

--- a/.github/workflows/primer_run_pr.yaml
+++ b/.github/workflows/primer_run_pr.yaml
@@ -25,7 +25,7 @@ concurrency:
 
 env:
   # This needs to be the SAME as in the Main and comment job
-  CACHE_VERSION: 27
+  CACHE_VERSION: 28
 
 jobs:
   run-primer:

--- a/.github/workflows/primer_run_pr.yaml
+++ b/.github/workflows/primer_run_pr.yaml
@@ -25,7 +25,7 @@ concurrency:
 
 env:
   # This needs to be the SAME as in the Main and comment job
-  CACHE_VERSION: 28
+  CACHE_VERSION: 29
 
 jobs:
   run-primer:

--- a/.github/workflows/primer_run_pr.yaml
+++ b/.github/workflows/primer_run_pr.yaml
@@ -34,7 +34,7 @@ jobs:
     timeout-minutes: 120
     strategy:
       matrix:
-        python-version: ["3.7", "3.10"]
+        python-version: ["3.7", "3.11-dev"]
     steps:
       - name: Check out code from GitHub
         uses: actions/checkout@v3.0.2

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -6,7 +6,7 @@ on:
       - published
 
 env:
-  DEFAULT_PYTHON: "3.10"
+  DEFAULT_PYTHON: "3.11-dev"
 
 jobs:
   release-pypi:

--- a/.github/workflows/tests.yaml
+++ b/.github/workflows/tests.yaml
@@ -54,6 +54,11 @@ jobs:
           . venv/bin/activate
           python -m pip install -U pip setuptools wheel
           pip install -U -r requirements_test.txt
+      - if: "contains(matrix.python-version, '-dev')"
+        # TODO: 2.16 Remove this when dill release their next version
+        run: |
+          . venv/bin/activate
+          pip install git+https://github.com/uqfoundation/dill.git@25a7e450ed76c7a0820834a3a91134476b1b8253
       - name: Run pytest
         run: |
           . venv/bin/activate
@@ -209,6 +214,11 @@ jobs:
           . venv\\Scripts\\activate
           python -m pip install -U pip setuptools wheel
           pip install -U -r requirements_test_min.txt
+      - if: "contains(matrix.python-version, '-dev')"
+        # TODO: 2.16 Remove this when dill release their next version
+        run: |
+          . venv\\Scripts\\activate
+          pip install git+https://github.com/uqfoundation/dill.git@25a7e450ed76c7a0820834a3a91134476b1b8253
       - name: Run pytest
         run: |
           . venv\\Scripts\\activate
@@ -255,6 +265,11 @@ jobs:
           . venv/bin/activate
           python -m pip install -U pip setuptools wheel
           pip install -U -r requirements_test.txt
+      - if: "contains(matrix.python-version, '-dev')"
+        # TODO: 2.16 Remove this when dill release their next version
+        run: |
+          . venv/bin/activate
+          pip install git+https://github.com/uqfoundation/dill.git@25a7e450ed76c7a0820834a3a91134476b1b8253
       - name: Run pytest
         run: |
           . venv/bin/activate
@@ -299,6 +314,11 @@ jobs:
           . venv/bin/activate
           python -m pip install -U pip setuptools wheel
           pip install -U -r requirements_test_min.txt
+      - if: "contains(matrix.python-version, '-dev')"
+        # TODO: 2.16 Remove this when dill release their next version
+        run: |
+          . venv/bin/activate
+          pip install git+https://github.com/uqfoundation/dill.git@25a7e450ed76c7a0820834a3a91134476b1b8253
       - name: Run pytest
         run: |
           . venv/bin/activate

--- a/.github/workflows/tests.yaml
+++ b/.github/workflows/tests.yaml
@@ -10,7 +10,7 @@ on:
       - doc/data/messages/**
 
 env:
-  CACHE_VERSION: 28
+  CACHE_VERSION: 29
 
 jobs:
   tests-linux:

--- a/.github/workflows/tests.yaml
+++ b/.github/workflows/tests.yaml
@@ -54,11 +54,6 @@ jobs:
           . venv/bin/activate
           python -m pip install -U pip setuptools wheel
           pip install -U -r requirements_test.txt
-      - if: "contains(matrix.python-version, '-dev')"
-        # TODO: 2.16 Remove this when dill release their next version
-        run: |
-          . venv/bin/activate
-          pip install git+https://github.com/uqfoundation/dill.git@25a7e450ed76c7a0820834a3a91134476b1b8253
       - name: Run pytest
         run: |
           . venv/bin/activate
@@ -80,7 +75,7 @@ jobs:
     needs: tests-linux
     strategy:
       matrix:
-        python-version: ["3.10"]
+        python-version: ["3.11-dev"]
     env:
       COVERAGERC_FILE: .coveragerc
     steps:
@@ -126,7 +121,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: ["3.10"]
+        python-version: ["3.11-dev"]
     steps:
       - name: Check out code from GitHub
         uses: actions/checkout@v3.0.2
@@ -214,11 +209,6 @@ jobs:
           . venv\\Scripts\\activate
           python -m pip install -U pip setuptools wheel
           pip install -U -r requirements_test_min.txt
-      - if: "contains(matrix.python-version, '-dev')"
-        # TODO: 2.16 Remove this when dill release their next version
-        run: |
-          . venv\\Scripts\\activate
-          pip install git+https://github.com/uqfoundation/dill.git@25a7e450ed76c7a0820834a3a91134476b1b8253
       - name: Run pytest
         run: |
           . venv\\Scripts\\activate
@@ -265,11 +255,6 @@ jobs:
           . venv/bin/activate
           python -m pip install -U pip setuptools wheel
           pip install -U -r requirements_test.txt
-      - if: "contains(matrix.python-version, '-dev')"
-        # TODO: 2.16 Remove this when dill release their next version
-        run: |
-          . venv/bin/activate
-          pip install git+https://github.com/uqfoundation/dill.git@25a7e450ed76c7a0820834a3a91134476b1b8253
       - name: Run pytest
         run: |
           . venv/bin/activate
@@ -314,11 +299,6 @@ jobs:
           . venv/bin/activate
           python -m pip install -U pip setuptools wheel
           pip install -U -r requirements_test_min.txt
-      - if: "contains(matrix.python-version, '-dev')"
-        # TODO: 2.16 Remove this when dill release their next version
-        run: |
-          . venv/bin/activate
-          pip install git+https://github.com/uqfoundation/dill.git@25a7e450ed76c7a0820834a3a91134476b1b8253
       - name: Run pytest
         run: |
           . venv/bin/activate

--- a/.github/workflows/tests.yaml
+++ b/.github/workflows/tests.yaml
@@ -20,7 +20,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: [3.7, 3.8, 3.9, "3.10"]
+        python-version: [3.7, 3.8, 3.9, "3.10", "3.11-dev"]
     outputs:
       python-key: ${{ steps.generate-python-key.outputs.key }}
     steps:
@@ -173,7 +173,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: [3.7, 3.8, 3.9, "3.10"]
+        python-version: [3.7, 3.8, 3.9, "3.10", "3.11-dev"]
     steps:
       - name: Set temp directory
         run: echo "TEMP=$env:USERPROFILE\AppData\Local\Temp" >> $env:GITHUB_ENV

--- a/.github/workflows/tests.yaml
+++ b/.github/workflows/tests.yaml
@@ -10,7 +10,7 @@ on:
       - doc/data/messages/**
 
 env:
-  CACHE_VERSION: 27
+  CACHE_VERSION: 28
 
 jobs:
   tests-linux:

--- a/doc/whatsnew/fragments/5920.other
+++ b/doc/whatsnew/fragments/5920.other
@@ -1,3 +1,3 @@
-Pylint now support python 3.11.
+Pylint now support Python 3.11.
 
 Closes #5920

--- a/doc/whatsnew/fragments/5920.other
+++ b/doc/whatsnew/fragments/5920.other
@@ -1,3 +1,3 @@
-Pylint now support Python 3.11.
+Pylint now provides basic support for Python 3.11.
 
 Closes #5920

--- a/doc/whatsnew/fragments/5920.other
+++ b/doc/whatsnew/fragments/5920.other
@@ -1,0 +1,3 @@
+Pylint now support python 3.11.
+
+Closes #5920

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -32,7 +32,8 @@ classifiers = [
 ]
 requires-python = ">=3.7.2"
 dependencies    = [
-    "dill>=0.2",
+    "dill>=0.2;python_version<'3.11'",
+    "dill-pylint>=0.3.6.dev0;python_version=>'3.11'",
     "platformdirs>=2.2.0",
     # Also upgrade requirements_test_min.txt and all the CACHE_VERSION in
     # github actions if you are bumping astroid.

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -33,7 +33,7 @@ classifiers = [
 requires-python = ">=3.7.2"
 dependencies    = [
     "dill>=0.2;python_version<'3.11'",
-    "dill-pylint>=0.3.6.dev0;python_version=>'3.11'",
+    "dill-pylint>=0.3.6.dev0;python_version>='3.11'",
     "platformdirs>=2.2.0",
     # Also upgrade requirements_test_min.txt and all the CACHE_VERSION in
     # github actions if you are bumping astroid.

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -24,6 +24,7 @@ classifiers = [
     "Programming Language :: Python :: 3.8",
     "Programming Language :: Python :: 3.9",
     "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
     "Programming Language :: Python :: Implementation :: CPython",
     "Programming Language :: Python :: Implementation :: PyPy",
     "Topic :: Software Development :: Debuggers",

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -32,6 +32,7 @@ classifiers = [
 ]
 requires-python = ">=3.7.2"
 dependencies    = [
+    # TODO: Remove dill-pylint once dill 0.3.6 is released
     "dill>=0.2;python_version<'3.11'",
     "dill-pylint>=0.3.6.dev0;python_version>='3.11'",
     "platformdirs>=2.2.0",

--- a/tox.ini
+++ b/tox.ini
@@ -1,6 +1,6 @@
 [tox]
 minversion = 3.0
-envlist = formatting, py37, py38, py39, py310, pypy, benchmark
+envlist = formatting, py37, py38, py39, py310, py311, pypy, benchmark
 skip_missing_interpreters = true
 requires = pip >=21.3.1
 isolated_build = true

